editor/components/rich-text/index.js

```diff
@@ -789,12 +789,33 @@ export class RichText extends Component {
 						return;
 					}
 
-					const anchor = this.editor.dom.getParent( this.editor.selection.getNode(), 'a' );
-					if ( ! anchor ) {
-						this.removeFormat( 'link' );
-					}
-					const { value: href, ...params } = formatValue;
-					this.applyFormat( 'link', { href, ...params }, anchor );
+					let anchor;
+
+					// Create only one undo level when inserting and modifying the link
+					this.editor.undoManager.transact( () => {
+						// Use built-in TinyMCE command to insert the link. This takes care of
+						// deleting any existing links within the selection
+						const { value: href, ...params } = formatValue;
+						this.editor.execCommand( 'mceInsertLink', false, {
+							href,
+							...params,
+							'data-wp-temp-link': 1,
+						} );
+
+						// mceInsertLink doesn't return the anchor it creates, so query for it
+						// using a temporary attribute
```

Comment:
I stole this idea from the classic editor.

https://github.com/WordPress/WordPress/blob/48077799d9d84b41d7ee1ea3a0054f0641813359/wp-includes/js/wplink.js#L418